File: web/src/lib/connectors.ts
import type { BoardObject, Side } from './types'

const SIDES: Side[] = ['top', 'right', 'bottom', 'left']
export const STUB = 24

export type Pt = { x: number; y: number }

export type Port = { side: Side; offset: number }

export type RouteOpts = {
  fromOffset?: number
  toOffset?: number
  waypoints?: number[]
  stub?: number
}

/** Local AABB so this module does not import ops (avoids cycles). */
function shapeBox(obj: BoardObject) {
  if ((obj.type === 'line' || obj.type === 'spline') && obj.points?.length) {
    let minX = Infinity
    let minY = Infinity
    let maxX = -Infinity
    let maxY = -Infinity
    for (let i = 0; i + 1 < obj.points.length; i += 2) {
      const px = obj.x + obj.points[i]
      const py = obj.y + obj.points[i + 1]
      minX = Math.min(minX, px)
      minY = Math.min(minY, py)
      maxX = Math.max(maxX, px)
      maxY = Math.max(maxY, py)
    }
    if (Number.isFinite(minX)) {
      return { x: minX, y: minY, w: Math.max(1, maxX - minX), h: Math.max(1, maxY - minY) }
    }
  }
  return { x: obj.x, y: obj.y, w: obj.w, h: obj.h }
}

function clamp01(t: number) {
  if (Number.isNaN(t)) return 0.5
  return Math.min(1, Math.max(0, t))
}

export function sideAnchor(obj: BoardObject, side: Side, offset = 0.5): Pt {
  const b = shapeBox(obj)
  const t = clamp01(offset)
  if (side === 'top') return { x: b.x + b.w * t, y: b.y }
  if (side === 'bottom') return { x: b.x + b.w * t, y: b.y + b.h }
  if (side === 'left') return { x: b.x, y: b.y + b.h * t }
  return { x: b.x + b.w, y: b.y + b.h * t }
}

export function sideNormal(side: Side): Pt {
  if (side === 'top') return { x: 0, y: -1 }
  if (side === 'bottom') return { x: 0, y: 1 }
  if (side === 'left') return { x: -1, y: 0 }
  return { x: 1, y: 0 }
}

export function nearestSides(from: BoardObject, to: BoardObject) {
  let best = { fromSide: 'right' as Side, toSide: 'left' as Side, dist: Infinity }
  for (const fs of SIDES) {
    const a = sideAnchor(from, fs)
    for (const ts of SIDES) {
      const b = sideAnchor(to, ts)
      const d = Math.hypot(a.x - b.x, a.y - b.y)
      if (d < best.dist) best = { fromSide: fs, toSide: ts, dist: d }
    }
  }
  return { fromSide: best.fromSide, toSide: best.toSide }
}

export function nearestPort(obj: BoardObject, point: Pt): Port {
  const b = shapeBox(obj)
  let best: Port = { side: 'right', offset: 0.5 }
  let bestDist = Infinity
  for (const side of SIDES) {
    let offset = 0.5
    if (side === 'top' || side === 'bottom') {
      offset = b.w > 0 ? (point.x - b.x) / b.w : 0.5
    } else {
      offset = b.h > 0 ? (point.y - b.y) / b.h : 0.5
    }
    offset = clamp01(offset)
    const a = sideAnchor(obj, side, offset)
    const d = Math.hypot(a.x - point.x, a.y - point.y)
    if (d < bestDist) {
      bestDist = d
      best = { side, offset }
    }
  }
  return best
}

function almostEq(a: number, b: number, eps = 0.5) {
  return Math.abs(a - b) < eps
}

/** Collapse consecutive duplicates and collinear middle points. */
export function simplifyOrthogonal(pts: Pt[]): Pt[] {
  if (pts.length <= 2) return pts.slice()
  const out: Pt[] = [{ ...pts[0] }]
  for (let i = 1; i < pts.length; i++) {
    const p = pts[i]
    const last = out[out.length - 1]
    if (almostEq(last.x, p.x) && almostEq(last.y, p.y)) continue
    out.push({ ...p })
  }
  let i = 1
  while (i < out.length - 1) {
    const a = out[i - 1]
    const b = out[i]
    const c = out[i + 1]
    const colH = almostEq(a.y, b.y) && almostEq(b.y, c.y)
    const colV = almostEq(a.x, b.x) && almostEq(b.x, c.x)
    if (colH || colV) out.splice(i, 1)
    else i++
  }
  return out
}

function flat(pts: Pt[]): number[] {
  const out: number[] = []
  for (const p of pts) out.push(p.x, p.y)
  return out
}

function parseWaypoints(waypoints?: number[]): Pt[] {
  if (!waypoints?.length) return []
  const out: Pt[] = []
  for (let i = 0; i + 1 < waypoints.length; i += 2) {
    out.push({ x: waypoints[i], y: waypoints[i + 1] })
  }
  return out
}

/** Auto orthogonal path from stubA to stubB (both already outside shapes). */
function routeBetweenStubs(a: Pt, fromSide: Side, b: Pt, toSide: Side): Pt[] {
  const fromH = fromSide === 'left' || fromSide === 'right'
  const toH = toSide === 'left' || toSide === 'right'

  if (fromH && toH) {
    // Both horizontal exits
    if ((fromSide === 'right' && a.x <= b.x) || (fromSide === 'left' && a.x >= b.x)) {
      if (almostEq(a.y, b.y)) return [a, { x: b.x, y: a.y }, b]
      const midX = (a.x + b.x) / 2
      return [a, { x: midX, y: a.y }, { x: midX, y: b.y }, b]
    }
    // Facing away or overlapping — go out then U-turn
    const midY = (a.y + b.y) / 2
    return [a, { x: a.x, y: midY }, { x: b.x, y: midY }, b]
  }

  if (!fromH && !toH) {
    // Both vertical exits
    if ((fromSide === 'bottom' && a.y <= b.y) || (fromSide === 'top' && a.y >= b.y)) {
      if (almostEq(a.x, b.x)) return [a, { x: a.x, y: b.y }, b]
      const midY = (a.y + b.y) / 2
      return [a, { x: a.x, y: midY }, { x: b.x, y: midY }, b]
    }
    const midX = (a.x + b.x) / 2
    return [a, { x: midX, y: a.y }, { x: midX, y: b.y }, b]
  }

  // Mixed: L or Z
  if (fromH) {
    // leave horizontally, arrive vertically
    const corner = { x: b.x, y: a.y }
    // If corner would reverse into target stub awkwardly, use Z
    if (
      (toSide === 'top' && corner.y > b.y) ||
      (toSide === 'bottom' && corner.y < b.y) ||
      (fromSide === 'right' && corner.x < a.x) ||
      (fromSide === 'left' && corner.x > a.x)
    ) {
      return [a, { x: b.x, y: a.y }, b]
    }
    return [a, corner, b]
  }

  // leave vertically, arrive horizontally
  const corner = { x: a.x, y: b.y }
  if (
    (toSide === 'left' && corner.x > b.x) ||
    (toSide === 'right' && corner.x < b.x) ||
    (fromSide === 'bottom' && corner.y < a.y) ||
    (fromSide === 'top' && corner.y > a.y)
  ) {
    return [a, { x: a.x, y: b.y }, b]
  }
  return [a, corner, b]
}

function enforceOrthogonal(pts: Pt[]): Pt[] {
  if (pts.length <= 1) return pts.slice()
  const out: Pt[] = [{ ...pts[0] }]
  for (let i = 1; i < pts.length; i++) {
    const prev = out[out.length - 1]
    const cur = pts[i]
    if (almostEq(prev.x, cur.x) || almostEq(prev.y, cur.y)) {
      out.push({ ...cur })
      continue
    }
    // Insert elbow: prefer matching previous segment direction if any
    out.push({ x: cur.x, y: prev.y })
    out.push({ ...cur })
  }
  return simplifyOrthogonal(out)
}

/**
 * Build full orthogonal polyline including anchors.
 * `waypoints` are interior corners only (absolute board coords).
 */
export function orthogonalRoute(
  from: BoardObject,
  to: BoardObject,
  fromSide: Side,
  toSide: Side,
  opts: RouteOpts = {},
): number[] {
  const stub = opts.stub ?? STUB
  const fromOff = opts.fromOffset ?? 0.5
  const toOff = opts.toOffset ?? 0.5
  const start = sideAnchor(from, fromSide, fromOff)
  const end = sideAnchor(to, toSide, toOff)
  const n0 = sideNormal(fromSide)
  const n1 = sideNormal(toSide)
  const stubA = { x: start.x + n0.x * stub, y: start.y + n0.y * stub }
  const stubB = { x: end.x + n1.x * stub, y: end.y + n1.y * stub }

  const custom = parseWaypoints(opts.waypoints)
  let middle: Pt[]
  if (custom.length) {
    middle = enforceOrthogonal([stubA, ...custom, stubB])
  } else {
    middle = simplifyOrthogonal(routeBetweenStubs(stubA, fromSide, stubB, toSide))
  }

  const full = simplifyOrthogonal([start, ...middle, end])
  return flat(full)
}

/** @deprecated Prefer orthogonalRoute; kept for simple 2-point callers. */
export function connectorPoints(
  from: BoardObject,
  to: BoardObject,
  fromSide: Side,
  toSide: Side,
  opts?: RouteOpts,
) {
  return orthogonalRoute(from, to, fromSide, toSide, opts)
}

export function connectorBounds(pts: number[]) {
  let minX = Infinity
  let minY = Infinity
  let maxX = -Infinity
  let maxY = -Infinity
  for (let i = 0; i + 1 < pts.length; i += 2) {
    minX = Math.min(minX, pts[i])
    minY = Math.min(minY, pts[i + 1])
    maxX = Math.max(maxX, pts[i])
    maxY = Math.max(maxY, pts[i + 1])
  }
  if (!Number.isFinite(minX)) return { x: 0, y: 0, w: 1, h: 1 }
  return {
    x: minX,
    y: minY,
    w: Math.max(1, maxX - minX),
    h: Math.max(1, maxY - minY),
  }
}

export function polylineMidpoint(pts: number[]): Pt {
  if (pts.length < 4) return { x: pts[0] ?? 0, y: pts[1] ?? 0 }
  let total = 0
  const segs: number[] = []
  for (let i = 0; i + 3 < pts.length; i += 2) {
    const len = Math.hypot(pts[i + 2] - pts[i], pts[i + 3] - pts[i + 1])
    segs.push(len)
    total += len
  }
  if (total < 1e-6) return { x: pts[0], y: pts[1] }
  let remain = total / 2
  for (let i = 0, s = 0; i + 3 < pts.length; i += 2, s++) {
    const len = segs[s]
    if (remain <= len) {
      const t = len < 1e-6 ? 0 : remain / len
      return {
        x: pts[i] + (pts[i + 2] - pts[i]) * t,
        y: pts[i + 1] + (pts[i + 3] - pts[i + 1]) * t,
      }
    }
    remain -= len
  }
  return { x: pts[pts.length - 2], y: pts[pts.length - 1] }
}

/** Arrowhead triangle at the end of the polyline. */
export function arrowHeadPoints(pts: number[], size = 10): number[] | null {
  if (pts.length < 4) return null
  const x2 = pts[pts.length - 2]
  const y2 = pts[pts.length - 1]
  let x1 = pts[pts.length - 4]
  let y1 = pts[pts.length - 3]
  // Walk back if last segment is tiny
  for (let i = pts.length - 4; i >= 0; i -= 2) {
    if (Math.hypot(x2 - pts[i], y2 - pts[i + 1]) > 1) {
      x1 = pts[i]
      y1 = pts[i + 1]
      break
    }
  }
  const dx = x2 - x1
  const dy = y2 - y1
  const len = Math.hypot(dx, dy) || 1
  const ux = dx / len
  const uy = dy / len
  const px = -uy
  const py = ux
  const baseX = x2 - ux * size
  const baseY = y2 - uy * size
  return [
    x2,
    y2,
    baseX + px * (size * 0.55),
    baseY + py * (size * 0.55),
    baseX - px * (size * 0.55),
    baseY - py * (size * 0.55),
  ]
}

/** Interior waypoints from a full route (drop first/last anchors and stubs if present). */
export function interiorWaypoints(route: number[]): number[] {
  const pts: Pt[] = []
  for (let i = 0; i + 1 < route.length; i += 2) pts.push({ x: route[i], y: route[i + 1] })
  if (pts.length <= 2) return []
  // Drop start and end anchors
  const inner = pts.slice(1, -1)
  return flat(inner)
}

/**
 * After dragging a segment between indices `segIndex` (point i → i+1 in route pts),
 * return new interior waypoints to persist.
 */
export function waypointsAfterSegmentDrag(
  route: number[],
  segIndex: number,
  axis: 'x' | 'y',
  value: number,
): number[] {
  const pts: Pt[] = []
  for (let i = 0; i + 1 < route.length; i += 2) pts.push({ x: route[i], y: route[i + 1] })
  if (segIndex < 0 || segIndex >= pts.length - 1) return interiorWaypoints(route)
  const a = pts[segIndex]
  const b = pts[segIndex + 1]
  if (axis === 'y') {
    a.y = value
    b.y = value
  } else {
    a.x = value
    b.x = value
  }
  // Keep anchors fixed: restore ends
  const start = { x: route[0], y: route[1] }
  const end = { x: route[route.length - 2], y: route[route.length - 1] }
  pts[0] = start
  pts[pts.length - 1] = end
  const ortho = enforceOrthogonal(pts)
  // Interior = everything except anchors
  if (ortho.length <= 2) return []
  return flat(ortho.slice(1, -1))
}

export function resolveEndpoint(
  objects: Record<string, BoardObject>,
  id: string | undefined,
  livePos?: Record<string, { x: number; y: number; w?: number; h?: number }>,
): BoardObject | undefined {
  if (!id) return undefined
  const o = objects[id]
  if (!o) return undefined
  const live = livePos?.[id]
  if (!live) return o
  return {
    ...o,
    x: live.x,
    y: live.y,
    ...(live.w !== undefined ? { w: live.w } : {}),
    ...(live.h !== undefined ? { h: live.h } : {}),
  }
}

export function routeForConnector(
  conn: BoardObject,
  objects: Record<string, BoardObject>,
  livePos?: Record<string, { x: number; y: number; w?: number; h?: number }>,
): number[] | null {
  const from = resolveEndpoint(objects, conn.fromId, livePos)
  const to = resolveEndpoint(objects, conn.toId, livePos)
  if (!from || !to) return null
  return orthogonalRoute(from, to, conn.fromSide ?? 'right', conn.toSide ?? 'left', {
    fromOffset: conn.fromOffset ?? 0.5,
    toOffset: conn.toOffset ?? 0.5,
    waypoints: conn.points,
  })
}

export { SIDES }

File: internal/draw/connectors.go
package draw

import (
	"math"

	"liro/internal/model"
)

const ConnectorStub = 24.0

type Pt struct {
	X, Y float64
}

func clamp01(t float64) float64 {
	if math.IsNaN(t) {
		return 0.5
	}
	if t < 0 {
		return 0
	}
	if t > 1 {
		return 1
	}
	return t
}

func almostEq(a, b, eps float64) bool {
	return math.Abs(a-b) < eps
}

// SideAnchorAt returns the point on the given side at offset t in [0,1].
func SideAnchorAt(o model.Object, side Side, offset float64) (x, y float64) {
	b := AABB(o)
	t := clamp01(offset)
	switch side {
	case SideTop:
		return b.X + b.W*t, b.Y
	case SideBottom:
		return b.X + b.W*t, b.Y + b.H
	case SideLeft:
		return b.X, b.Y + b.H*t
	default:
		return b.X + b.W, b.Y + b.H*t
	}
}

func SideNormal(side Side) Pt {
	switch side {
	case SideTop:
		return Pt{0, -1}
	case SideBottom:
		return Pt{0, 1}
	case SideLeft:
		return Pt{-1, 0}
	default:
		return Pt{1, 0}
	}
}

func simplifyOrthogonal(pts []Pt) []Pt {
	if len(pts) <= 2 {
		out := make([]Pt, len(pts))
		copy(out, pts)
		return out
	}
	out := []Pt{pts[0]}
	for i := 1; i < len(pts); i++ {
		p := pts[i]
		last := out[len(out)-1]
		if almostEq(last.X, p.X, 0.5) && almostEq(last.Y, p.Y, 0.5) {
			continue
		}
		out = append(out, p)
	}
	i := 1
	for i < len(out)-1 {
		a, b, c := out[i-1], out[i], out[i+1]
		colH := almostEq(a.Y, b.Y, 0.5) && almostEq(b.Y, c.Y, 0.5)
		colV := almostEq(a.X, b.X, 0.5) && almostEq(b.X, c.X, 0.5)
		if colH || colV {
			out = append(out[:i], out[i+1:]...)
			continue
		}
		i++
	}
	return out
}

func routeBetweenStubs(a Pt, fromSide Side, b Pt, toSide Side) []Pt {
	fromH := fromSide == SideLeft || fromSide == SideRight
	toH := toSide == SideLeft || toSide == SideRight

	if fromH && toH {
		if (fromSide == SideRight && a.X <= b.X) || (fromSide == SideLeft && a.X >= b.X) {
			if almostEq(a.Y, b.Y, 0.5) {
				return []Pt{a, {b.X, a.Y}, b}
			}
			midX := (a.X + b.X) / 2
			return []Pt{a, {midX, a.Y}, {midX, b.Y}, b}
		}
		midY := (a.Y + b.Y) / 2
		return []Pt{a, {a.X, midY}, {b.X, midY}, b}
	}

	if !fromH && !toH {
		if (fromSide == SideBottom && a.Y <= b.Y) || (fromSide == SideTop && a.Y >= b.Y) {
			if almostEq(a.X, b.X, 0.5) {
				return []Pt{a, {a.X, b.Y}, b}
			}
			midY := (a.Y + b.Y) / 2
			return []Pt{a, {a.X, midY}, {b.X, midY}, b}
		}
		midX := (a.X + b.X) / 2
		return []Pt{a, {midX, a.Y}, {midX, b.Y}, b}
	}

	if fromH {
		return []Pt{a, {b.X, a.Y}, b}
	}
	return []Pt{a, {a.X, b.Y}, b}
}

func enforceOrthogonal(pts []Pt) []Pt {
	if len(pts) <= 1 {
		out := make([]Pt, len(pts))
		copy(out, pts)
		return out
	}
	out := []Pt{pts[0]}
	for i := 1; i < len(pts); i++ {
		prev := out[len(out)-1]
		cur := pts[i]
		if almostEq(prev.X, cur.X, 0.5) || almostEq(prev.Y, cur.Y, 0.5) {
			out = append(out, cur)
			continue
		}
		out = append(out, Pt{cur.X, prev.Y}, cur)
	}
	return simplifyOrthogonal(out)
}

func flatPts(pts []Pt) []float64 {
	out := make([]float64, 0, len(pts)*2)
	for _, p := range pts {
		out = append(out, p.X, p.Y)
	}
	return out
}

func parseWaypoints(waypoints []float64) []Pt {
	out := make([]Pt, 0, len(waypoints)/2)
	for i := 0; i+1 < len(waypoints); i += 2 {
		out = append(out, Pt{waypoints[i], waypoints[i+1]})
	}
	return out
}

// OrthogonalRoute builds a Manhattan polyline between two objects.
// waypoints are interior corners only; empty means fully automatic.
func OrthogonalRoute(from, to model.Object, fromSide, toSide Side, fromOffset, toOffset float64, waypoints []float64, stub float64) []float64 {
	if stub <= 0 {
		stub = ConnectorStub
	}
	sx, sy := SideAnchorAt(from, fromSide, fromOffset)
	ex, ey := SideAnchorAt(to, toSide, toOffset)
	n0 := SideNormal(fromSide)
	n1 := SideNormal(toSide)
	stubA := Pt{sx + n0.X*stub, sy + n0.Y*stub}
	stubB := Pt{ex + n1.X*stub, ey + n1.Y*stub}

	var middle []Pt
	custom := parseWaypoints(waypoints)
	if len(custom) > 0 {
		chain := append([]Pt{stubA}, custom...)
		chain = append(chain, stubB)
		middle = enforceOrthogonal(chain)
	} else {
		middle = simplifyOrthogonal(routeBetweenStubs(stubA, fromSide, stubB, toSide))
	}

	full := []Pt{{sx, sy}}
	full = append(full, middle...)
	full = append(full, Pt{ex, ey})
	return flatPts(simplifyOrthogonal(full))
}

// ConnectorRoute resolves sides/offsets from a connector object and routes it.
func ConnectorRoute(conn model.Object, objects map[string]model.Object) []float64 {
	from, ok1 := objects[conn.FromID]
	to, ok2 := objects[conn.ToID]
	if !ok1 || !ok2 {
		return nil
	}
	fs, okf := ParseSide(conn.FromSide)
	ts, okt := ParseSide(conn.ToSide)
	if !okf || !okt {
		nfs, nts := NearestSides(from, to)
		if !okf {
			fs = nfs
		}
		if !okt {
			ts = nts
		}
	}
	fromOff, toOff := 0.5, 0.5
	if conn.FromOffset != nil {
		fromOff = clamp01(*conn.FromOffset)
	}
	if conn.ToOffset != nil {
		toOff = clamp01(*conn.ToOffset)
	}
	return OrthogonalRoute(from, to, fs, ts, fromOff, toOff, conn.Points, ConnectorStub)
}

// BoundsOfPoints returns the AABB of a flat polyline.
func BoundsOfPoints(pts []float64) Box {
	if len(pts) < 2 {
		return Box{0, 0, 1, 1}
	}
	minX, minY := math.Inf(1), math.Inf(1)
	maxX, maxY := math.Inf(-1), math.Inf(-1)
	for i := 0; i+1 < len(pts); i += 2 {
		if pts[i] < minX {
			minX = pts[i]
		}
		if pts[i+1] < minY {
			minY = pts[i+1]
		}
		if pts[i] > maxX {
			maxX = pts[i]
		}
		if pts[i+1] > maxY {
			maxY = pts[i+1]
		}
	}
	return Box{minX, minY, math.Max(1, maxX-minX), math.Max(1, maxY-minY)}
}
